apps/omg_api/lib/state/core.ex

```diff
@@ -168,98 +163,111 @@ defmodule OMG.API.State.Core do
-         spender
+  defp inputs_belong_to_spenders?(
+         utxos,
+         %Transaction.Recovered{
+           signed_tx: %Transaction.Signed{raw_tx: raw_tx}
+         } = recovered_tx
        ) do
-    with :ok <- utxo_not_from_the_future_block?(state, blknum) do
-      check_utxo_and_extract_amount(state, Utxo.position(blknum, txindex, oindex), spender, spent_cur)
+    inputs = Transaction.get_inputs(raw_tx)
+
+    with {:ok, input_utxos} <- get_input_utxos(utxos, [], inputs),
+         input_utxos_owners <- Enum.map(input_utxos, fn %{owner: owner} -> owner end),
+         :ok <- Transaction.Recovered.all_spenders_authorized?(recovered_tx, input_utxos_owners) do
+      {:ok, input_utxos}
     end
   end
 
-  defp check_utxo_and_extract_amount(%Core{utxos: utxos}, position, spender, spent_cur) do
-    with {:ok, %Utxo{owner: owner, currency: cur, amount: owner_has}} <- get_utxo(utxos, position),
-         :ok <- is_spender?(owner, spender),
-         :ok <- same_currency?(cur, spent_cur),
-         do: {:ok, owner_has}
-  end
+  defp get_input_utxos(_, acc, []), do: {:ok, acc}
 
-  defp utxo_not_from_the_future_block?(%__MODULE__{height: blknum}, input_blknum) do
-    if blknum >= input_blknum do
-      :ok
-    else
-      {:error, :input_utxo_ahead_of_state}
-    end
-  end
+  defp get_input_utxos(utxos, acc, [%{blknum: 0, txindex: 0, oindex: 0} | inputs]),
+    do: get_input_utxos(utxos, acc, inputs)
+
+  defp get_input_utxos(utxos, acc, [%{blknum: blknum, txindex: txindex, oindex: oindex} | inputs]) do
+    position = Utxo.position(blknum, txindex, oindex)
 
-  defp get_utxo(utxos, position) do
     case Map.get(utxos, position) do
       nil -> {:error, :utxo_not_found}
-      found -> {:ok, found}
+      found -> get_input_utxos(utxos, [found | acc], inputs)
     end
   end
 
-  defp is_spender?(owner, spender) do
-    if owner == spender, do: :ok, else: {:error, :incorrect_spender}
+  defp get_amounts_by_currency(utxos) do
+    utxos
+    |> Enum.group_by(fn %{currency: currency} -> currency end, fn %{amount: amount} -> amount end)
+    |> Enum.map(fn {currency, amounts} -> {currency, Enum.sum(amounts)} end)
+    |> Map.new()
+  end
+
+  # fee is implicit - it's the difference between funds owned and spend
+  defp amounts_add_up?(input_amounts, output_amounts, fees) do
+    with :ok <- same_currencies?(input_amounts, output_amounts) do
+      amounts_add_up =
+        for {input_currency, input_amount} <- Map.to_list(input_amounts) do
+          fee = Map.get(fees, input_currency)
+          output_amount = Map.get(output_amounts, input_currency)
+          input_amount >= fee + output_amount
+        end
+        |> Enum.all?()
+
+      if amounts_add_up, do: :ok, else: {:error, :amounts_do_not_add_up}
+    end
   end
 
-  defp same_currency?(utxo_currency, spent_currency) do
-    if utxo_currency == spent_currency, do: :ok, else: {:error, :incorrect_currency}
+  defp same_currencies?(input_amounts, output_amounts) do
```

Comment:
probably not necessary and would hurt in later developments - what if someone adds an input in a currency solely to cover a tx fee?

Maybe it's just safer to do `amounts_do_not_add_up` in situations when the output currencies aren't covered by inputs.

Reply:
Sorry to jump into the conversation.

When we first launch, it's likely we will only accept a very restricted list of currencies for fees (maybe only ETH to begin with). So it's very likely that if a user is sending an ERC20 token, that they will need to add an input to pay the fee.